File: app/src/main/java/org/miaowo/miaowo/fragment/chat/ChatListFragment.java
package org.miaowo.miaowo.fragment.chat;

import android.content.Context;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import org.miaowo.miaowo.R;
import org.miaowo.miaowo.adapter.ChatRoomListAdapter;
import org.miaowo.miaowo.api.API;
import org.miaowo.miaowo.bean.data.ChatRoom;
import org.miaowo.miaowo.bean.data.ChatRoomList;
import org.miaowo.miaowo.custom.load_more_list.LMLPageAdapter;
import org.miaowo.miaowo.custom.load_more_list.LoadMoreList;
import org.miaowo.miaowo.root.BaseActivity;
import org.miaowo.miaowo.root.BaseFragment;
import org.miaowo.miaowo.util.HttpUtil;
import org.miaowo.miaowo.util.JsonUtil;

import java.io.IOException;

import okhttp3.Request;

public class ChatListFragment extends BaseFragment {

    private OnChatListener mListener;
    private LoadMoreList mList;
    private LMLPageAdapter<ChatRoom> mAdapter;
    private HttpUtil mHttp;
    private JsonUtil mJson;
    public ChatListFragment() {}

    @SuppressWarnings("unused")
    public static ChatListFragment newInstance() {
        return new ChatListFragment();
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return new LoadMoreList(getContext());
    }

    @Override
    public void initView(View view) {
        if (view instanceof LoadMoreList) {
            mList = (LoadMoreList) view;
            mHttp = HttpUtil.utils();
            mJson = JsonUtil.utils();

            mAdapter = new ChatRoomListAdapter(getContext(), mListener);
            mList.setAdapter(mAdapter);
            mList.setPullRefresher(this::loadChatList, false);
            mList.load();
        }
    }

    private void loadChatList() {
        mList.loadMoreControl(false, false);
        Request request = new Request.Builder().url(getActivity().getString(R.string.url_chat_room, API.loginUser.getUsername().toLowerCase())).build();
        mHttp.post(request, (call, response) -> BaseActivity.get.runOnUiThreadIgnoreError(() -> {
                    try {
                        mAdapter.update(mJson.buildFromAPI(response, ChatRoomList.class).getRooms());
                        mList.loadMoreControl(true, true);
                        mList.loadOver();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }));
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnChatListener) {
            mListener = (OnChatListener) context;
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    public interface OnChatListener {
        void toRoom(ChatRoom room);
    }
}

File: app/src/main/java/org/miaowo/miaowo/fragment/chat/ChatListFragment.kt
package org.miaowo.miaowo.fragment.chat

import android.os.Bundle
import android.support.v7.widget.RecyclerView
import android.view.View
import org.miaowo.miaowo.API
import org.miaowo.miaowo.Miao
import org.miaowo.miaowo.R
import org.miaowo.miaowo.adapter.ChatRoomListAdapter
import org.miaowo.miaowo.base.extra.*
import org.miaowo.miaowo.bean.data.User
import org.miaowo.miaowo.fragment.user.UserListFragment
import org.miaowo.miaowo.other.Const
import org.miaowo.miaowo.other.MiaoListFragment

class ChatListFragment : MiaoListFragment(R.string.chat, true) {
    companion object {
        fun newInstance(): ChatListFragment {
            val fragment = ChatListFragment()
            val args = Bundle()
            args.putBoolean(Const.FG_POP_ALL, true)
            args.putString(Const.TAG, "${fragment.javaClass.name}.user.${API.user.uid}")
            fragment.arguments = args
            return fragment
        }
    }

    private val mAdapter = ChatRoomListAdapter()

    override fun setAdapter(list: RecyclerView) {
        list.adapter = mAdapter
    }

    override fun onRefresh() {
        API.Doc.chatRoom {
            mAdapter.update(it?.rooms ?: listOf())
            super.onRefresh()
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        miaoListener?.button?.setOnClickListener {
            val fg = UserListFragment.newInstance()
            val call = object : FragmentCall("newRoom", this, fg) {
                override fun call(vararg params: Any?) {
                    newRoom(params[0] as User)
                }
            }
            registerCall(call)
            fg.showSelf(Miao.i, this)
        }
    }

    override fun onClickListener(view: View, position: Int): Boolean {
        val clickItem = mAdapter.getItem(position)
        Miao.i.showFragment(ChatFragment.newInstance(clickItem), this)
        return true
    }

    private fun newRoom(user: User) = lInfo(user)
}